cpu/stm32_common/periph/gpio.c

```diff
@@ -163,6 +171,10 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
 
 void gpio_init_af(gpio_t pin, gpio_af_t af)
 {
+    if (pin == GPIO_UNDEF) {
```

Comment:
We decided at some point that periph functions should only check validity of their device/pin parameters in init() functions of the public API.
Thus please change the checks everywhere but in ```gpio_init()``` and ```gpio_init_int()``` to asserts.

Reply:
It makes GPIO_UNDEF totally meaningless as it can not be used in any real-world code then in any way except checking it before ever gpio_* call manually.

Reply:
> It makes GPIO_UNDEF totally meaningless as it can not be used in any real-world code then in any way except checking it before ever gpio_* call manually.

Please stop judging what is totally meaningless and what can be used in real-world code, and stay objective.

The rationale is that parameters like gpio_t have to be passed to e.g., ```gpio_init()``` anyways. Using gpio_set() on a pin not initialized with gpio_init() leads to undefined behaviour.
If gpio_set() would *always* check, it would increase codesize and reduze performance. Without the check, a gpio_set() optimized, with LTO, from a function call down to one wor two instructions, in the best case.
Checking the return code of gpio_set() wherever it is used would even more bloat the code.
Probably, noone would do it. And silently returning (instead of crashing early during development by assert) would leave to undefined behaviour in different ways. We chose the less bloated, high performance way of failing.
Mind you, gpio is a very low level interface. It might even be used to directly drive clock signals for drivers. The extra checking is overkill there.

If you still disagree, open an issue.

Reply:
> Please stop judging what is totally meaningless and what can be used in real-world code, and stay objective.

Objective: almost any RIOT driver in the release version of this fine OS which uses GPIO will crash the whole system immediately if supplied with GPIO_UNDEF.

Is this a normal situation?

Go on. Rewrite the drivers. Rewrite the code. Put a statement on GPIO_UNDEF that it must not be supplied to anything expect gpio_init. Add a statement that every gpio_set or gpio_read call must check if that gpio was properly initialized, even if gpio_init_int was miles away in different routine in a different C file.

Let's take periph_common/spi.c, I see, you wrote it.

'''
if ((cs == SPI_CS_UNDEF) || (cs == GPIO_UNDEF)) {
        return SPI_NOCS;
    }
'''

Where exactly in your own code it is checked that *cs* was properly initialized?

It is not.

Will you fix it?

> It might even be used to directly drive clock signals for drivers

Any *real* use case, as you like to say? With over 1 MHz GPIO bitbanging, ruined by a few CPU cycles added to the routine? No? None I expected.

And here is the perfectly real use case: two boards, one of them has two LEDs, another has one. And your proposition is to add explicit check of BOARD_LED_2 value before every LED blink in every piece of code where it should blink instead of a simple single check in gpio_set?

Elegant. Simple. No code duplication at all.

> If you still disagree, open an issue.

Seriously, do I look like I care?

I show you the problem that crashes your fine OS with a perfectly legitimate code. Will you fix it or not — not my problem, actually. It is yours. So I'm just not gonna spend another week on pointless discussion.

Reply:
> I show you the problem that crashes your fine OS with a perfectly legitimate code.

Supplying ```gpio_read/write()``` with it's first parameter not successfully initialized using ```gpio_init(_int)``` is not perfectly legitimate code.

Anyhow, that's off-topic. For this PR, please change the mentioned checks (all but init/init_int()) to assert.

BTW, #8040 shows 20% overhead for this PR's changes.

Reply:
> Anyhow, that's off-topic. For this PR, please change the mentioned checks

I would not.

It seems you think I am somehow interested in pushing this PR to mainline.

Wrong assumption. Why should I? I am not interested in working on faulty, error-prone concepts because maintainer who can't get his own code straight wants it that way.

You want your OS to be fixed on your indisputable terms — ok, so you go and fix it.